gk/fib.c

```diff
@@ -0,0 +1,1640 @@
+
+	/*
+	 * The @ip_prefix is a gateway.
+	 * Find the FIB entry for the gateway.
+	 * We need to release the @eth_cache
+	 * Ethernet header entry from the neighbor hash table.
+	 */
+	if (gw_addr->proto == ETHER_TYPE_IPv4 && gw_addr->ip.v4.s_addr
+			== ip_prefix_addr->ip.v4.s_addr) {
+		/*
+		 * The Ethernet header cache entry is still
+		 * referenced by another prefix, so we cannot
+		 * release this FIB entry.
+		 */
+		if (eth_cache->ref_cnt != 1) {
+			RTE_LOG(ERR, GATEKEEPER,
+				"gk: delete a gateway fib entry [ip prefix = %s], however, its cached Ethernet header is referenced by other entries\n",
+				ip_prefix);
+			return -1;
+		}
+
+		if (ip_prefix_fib->action == GK_FWD_GATEWAY_FRONT_NET)
+			neigh_fib = gk_conf->neigh_fib_front;
+		else
+			neigh_fib = gk_conf->neigh_fib_back;
+
+		neigh_ht = &neigh_fib->u.neigh;
+		ret = neigh_del_ether_cache(neigh_ht,
+			&gw_addr->ip.v4.s_addr);
+		if (ret < 0) {
+			RTE_LOG(ERR, GATEKEEPER,
+				"gk: failed to delete the Ethernet cached header for the IP prefix %s\n",
+				ip_prefix);
+			return -1;
+		}
+	} else if (gw_addr->proto == ETHER_TYPE_IPv6 &&
+			memcmp(gw_addr->ip.v6.s6_addr,
+			ip_prefix_addr->ip.v6.s6_addr,
+			sizeof(gw_addr->ip.v6) == 0)) {
+		/*
+		 * The Ethernet header cache entry is still
+		 * referenced by another prefix, so we cannot
+		 * release this FIB entry.
+		 */
+		if (eth_cache->ref_cnt != 1) {
+			RTE_LOG(ERR, GATEKEEPER,
+				"gk: delete a gateway fib entry [ip prefix = %s], however, its cached Ethernet header is referenced by other entries\n",
+				ip_prefix);
+			return -1;
+		}
+
+		if (ip_prefix_fib->action == GK_FWD_GATEWAY_FRONT_NET)
+			neigh_fib = gk_conf->neigh6_fib_front;
+		else
+			neigh_fib = gk_conf->neigh6_fib_back;
+
+		neigh_ht = &neigh_fib->u.neigh6;
+		ret = neigh_del_ether_cache(neigh_ht,
+			gw_addr->ip.v6.s6_addr);
+		if (ret < 0) {
+			RTE_LOG(ERR, GATEKEEPER,
+				"gk: failed to delete the Ethernet cached header for the IP prefix %s\n",
+				ip_prefix);
+			return -1;
+		}
+	} else {
+		/*
+		 * The @ip_prefix is not a gateway, and it has
+		 * a reference to a gateway entry.
+		 * We need to decrement the @eth_cache->ref_cnt.
+		 */
+		eth_cache->ref_cnt--;
+	}
+
+	return 0;
+}
+
+/*
+ * For removing FIB entries, once the field @ref_cnt
+ * of a FIB entry becomes 0, then it needs to notify the GK instances
+ * to flush the flows that have a reference to it. This can be
+ * implemented by mailbox ring.
+ */
+static int
+del_fib_entry_locked(const char *ip_prefix, struct gk_config *gk_conf)
+{
+	int ret;
+	int ip_prefix_len;
+	struct ipaddr ip_prefix_addr;
+	struct gk_fib *ip_prefix_fib;
+
+	if (ip_prefix == NULL)
+		return -1;
+
+ 	ip_prefix_len = parse_ip_prefix(ip_prefix, &ip_prefix_addr);
+	if (ip_prefix_len < 0)
+		return -1;
+
+	ip_prefix_fib = check_fib_entry_for_del_locked(
+		ip_prefix, &ip_prefix_addr, ip_prefix_len, gk_conf);
+	if (ip_prefix_fib == NULL)
+		return -1;
+
+	switch (ip_prefix_fib->action) {
```

Comment:
We need to notify the GK instances that we are going to remove `ip_prefix_fib` prefix before we execute this `switch` because the functions after this `switch` are going to change the still potentially accessible FIB entry `ip_prefix_fib`. We can only edit and release `ip_prefix_fib` after the GK blocks don't have references to it.